### PR TITLE
Remove `bind` from `T.proc` when in Initializer rewriter

### PR DESCRIPTION
**Pull request by pl-at-shopify**
This little patch simply removes `.bind(…)` from the chain of `Send`s when copying the block argument type definition to the `initialize` method body where `.bind(…)` is not allowed.

### Motivation

Fixes https://github.com/sorbet/sorbet/issues/6848.

### Test plan

See included automated tests.

### CODE REVIEW

File: rewriter/Initializer.cc [hunk 1]
```diff
@@ -29,6 +29,30 @@ bool isCopyableType(const ast::ExpressionPtr &typeExpr) {
...
+void maybeRemoveBind(core::Context ctx, ast::Send *send) {
+    auto lastSend = send;
+    while (send != nullptr) {
+        if (send->fun == core::Names::bind()) {
+            // `bind` is the last send in the chain of sends.
+            // This is incorrect syntax which is covered by the resolver.
+            // Safe to ignore.
+            if (send == lastSend) {
+                return;
+            }
```
Comment (jez) [inline rewriter/Initializer.cc:42]: Do you mind writing a test for this case?

File: rewriter/Initializer.cc [hunk 2]
```diff
@@ -29,6 +29,30 @@ bool isCopyableType(const ast::ExpressionPtr &typeExpr) {
...
+            auto recvSend = ast::cast_tree<ast::Send>(send->recv);
+            if (recvSend != nullptr) {
+                lastSend->recv = move(send->recv);
+                return;
+            }
+            // We handle only one `.bind` call in the whole chain,
+            // no need to traverse the rest of the tree.
+            return;
```
Comment (jez) [inline rewriter/Initializer.cc:50]: And this as well?

**Review by jez — APPROVED** [review]
Looks good but I would appreciate two more tests.

**Comment by jez** [comment]
We have a policy of testing changes to Sorbet against Stripe's codebase before
merging them. I've kicked off a test run for the current PR.  When the build
finishes, I'll share with you whether or how it failed. Thanks!

Stripe employees can see the build results here:

→ https://go/builds/bui_NkWN0wTKEcTkq4
→ https://go/builds/bui_NkWNkY1hhH4kBp
→ https://go/builds/bui_NkWNhtrRCPSKqp

**Comment by jez** [comment]
> Looks good but I would appreciate two more tests.

Going to merge for now. Feel free to open a new PR with these tests.

File: rewriter/Initializer.cc (hunk 1 above)
Comment (pl-at-shopify) [reply rewriter/Initializer.cc:42]: Got into a sort of trouble testing it. The test requires marking the same line with two `# error:` comments and I'm not sure how to deal with it. Here is the test code:

```ruby
class TProcBindInInitializerLastSend
  extend T::Sig

  sig {params(blk: T.proc.bind(String)).void }
    @blk = blk
  end
end

```

The line `sig {params(blk: T.proc.bind(String)).void }` produces two errors now:
- Malformed T.proc: You must specify a return type
- Using `bind` is not permitted here

One way to get around it is to change the code in the PR, but it looks so clean and bullet proof that I'm hesitating to :)

File: rewriter/Initializer.cc (hunk 2 above)
Comment (pl-at-shopify) [reply rewriter/Initializer.cc:50]: Same as above with these two errors competing for the same line of code:
- Malformed `bind`: Multiple calls to `.bind`
- Using `bind` is not permitted here

Comment (froydnj) [reply rewriter/Initializer.cc:50]: We have docs on how to add multiple errors for a single line: https://github.com/sorbet/sorbet/#test_corpus-tests

**Comment by pl-at-shopify** [comment]
Here the tests go: https://github.com/sorbet/sorbet/pull/6955
Thanks for trusting me on them coming 👍

File: rewriter/Initializer.cc (hunk 2 above)
Comment (jez) [reply rewriter/Initializer.cc:50]: +1 to drive home what nathan mentions:

```
T.proc.bind(A).bind(A).void
#      ^^^^ error: first error
#              ^^^^ error: second error
```

should let you make these assertions

**Pull request merged**